### PR TITLE
Add flags to configure secure connection client certificate

### PR DESCRIPTION
**Pull request by gkoitchev-bosch**
[#10] Add flags to configure secure connection client certificate

There should be an option to configure loading client certificate from file, instead of using the system pool.

Signed-off-by: Georgi Koitchev <georgi.koitchev@bosch.io>

### CODE REVIEW

File: internal/flags.go [hunk 1]
```diff
@@ -37,6 +37,8 @@ const (
 	flagLogFileSize     = "logFileSize"
 	flagLogFileCount    = "logFileCount"
 	flagLogFileMaxAge   = "logFileMaxAge"
+	flagCert            = "cert"
```
Comment (e-grigorov) [inline internal/flags.go:40]: Is it correct to use such generic names? How are we going to add MQTTS later on? Overall concept has to be integrated. For example, these flags can be: downloadCert, downloadKey.

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: internal/flags.go [hunk 2]
```diff
@@ -58,45 +52,35 @@ var (
...
-	LogFileMaxAge   int      `json:"logFileMaxAge"`
+	Broker          string   `json:"broker" def:"tcp://localhost:1883" descr:"Local MQTT broker address"`
+	Username        string   `json:"username" descr:"Username for authorized local client"`
+	Password        string   `json:"password" descr:"Password for authorized local client"`
+	StorageLocation string   `json:"storageLocation" descr:"Location of the storage"`
+	FeatureID       string   `json:"featureId" def:"SoftwareUpdatable" descr:"Feature identifier of SoftwareUpdatable"`
+	ModuleType      string   `json:"moduleType" def:"software" descr:"Module type of SoftwareUpdatable"`
+	ArtifactType    string   `json:"artifactType" def:"archive" descr:"Defines the module artifact type: archive or plane"`
+	Install         []string `json:"install" descr:"Defines the absolute path to install script"`
+	Cert            string   `json:"cert" descr:"A PEM encoded certificate \"file\" for secure artifact download"`
```
Comment (gboyvalenkov-bosch) [inline internal/flags.go:63]: Rename to ServerCert

File: internal/flags.go [hunk 3]
```diff
@@ -58,45 +52,35 @@ var (
...
+	Key             string   `json:"key" descr:"A PEM encoded unencrypted private key for secure artifact download"`
```
Comment (gboyvalenkov-bosch) [inline internal/flags.go:64]: Remove this flag

File: internal/flags_test.go [hunk 4]
```diff
@@ -110,6 +111,8 @@ func TestFlagsHasHigherPriority(t *testing.T) {
 	expectedArtifact := "TestArchive"
 	expectedFeatureID := "TestFeature"
 	expectedInstall := "TestInstall"
+	expectedCert := "TestCert"
+	expectedKey := "TestKey"
```
Comment (gboyvalenkov-bosch) [inline internal/flags_test.go:115]: remove

File: internal/flags_test.go [hunk 5]
```diff
@@ -126,6 +129,8 @@ func TestFlagsHasHigherPriority(t *testing.T) {
 		c(flagArtifactType, expectedArtifact),
 		c(flagFeatureID, expectedFeatureID),
 		c(flagInstall, expectedInstall),
+		c(flagCert, expectedCert),
+		c(flagKey, expectedKey),
```
Comment (gboyvalenkov-bosch) [inline internal/flags_test.go:133]: remove

File: internal/flags_test.go [hunk 6]
```diff
@@ -147,6 +152,8 @@ func TestFlagsHasHigherPriority(t *testing.T) {
 		Broker:          expectedFlagBroker,
 		Username:        expectedUsername,
 		Password:        expectedPassword,
+		Cert:            expectedCert,
+		Key:             expectedKey,
```
Comment (gboyvalenkov-bosch) [inline internal/flags_test.go:156]: remove

File: internal/storage/download.go [hunk 7]
```diff
@@ -31,7 +35,7 @@ import (
 const prefix = "_temporary-"
 
 // downloadArtifact tries to resume previous download operation or perform a new download.
-func downloadArtifact(to string, artifact *Artifact, progress progressBytes, done chan struct{}) error {
+func downloadArtifact(to string, artifact *Artifact, progress progressBytes, cert, key string, done chan struct{}) error {
```
Comment (gboyvalenkov-bosch) [inline internal/storage/download.go:38]: key not needed

File: internal/storage/download.go [hunk 8]
```diff
@@ -143,6 +139,58 @@ func resume(to string, offset int64, artifact *Artifact, progress progressBytes,
 	return validate(to, artifact.HashType, artifact.HashValue)
 }
 
+func getDownloadResponse(link string, offset int64, certFile, keyFile string) (*http.Response, error) {
```
Comment (gboyvalenkov-bosch) [inline internal/storage/download.go:142]: consider renaming to requestDownload

File: licenses/LICENSE.eclipse_ditto_clients_golang-0.0.0-20211126080925-0676267c80ac [hunk 9]
```diff
@@ -0,0 +1,89 @@
+Eclipse Public License - v 2.0
```
Comment (gboyvalenkov-bosch) [inline licenses/LICENSE.eclipse_ditto_clients_golang-0.0.0-20211126080925-0676267c80ac:1]: remove this file

File: licenses/LICENSE.eclipse_paho_mqtt_clients_golang-1.3.5 [hunk 10]
```diff
@@ -0,0 +1,13 @@
+Eclipse Distribution License - v 1.0
```
Comment (gboyvalenkov-bosch) [inline licenses/LICENSE.eclipse_paho_mqtt_clients_golang-1.3.5:1]: remove this file

File: licenses/LICENSE.fsnotify_fsnotify-1.5.1 [hunk 11]
```diff
@@ -0,0 +1,28 @@
+Copyright (c) 2012 The Go Authors. All rights reserved.
```
Comment (gboyvalenkov-bosch) [inline licenses/LICENSE.fsnotify_fsnotify-1.5.1:1]: remove this file

File: licenses/LICENSE.golang_go-1.17.2 [hunk 12]
```diff
@@ -0,0 +1,27 @@
+Copyright (c) 2009 The Go Authors. All rights reserved.
```
Comment (gboyvalenkov-bosch) [inline licenses/LICENSE.golang_go-1.17.2:1]: remove this file

File: licenses/LICENSE.golang_org_x_net-0.0.0-20211123203042-d83791d6bcd9 [hunk 13]
```diff
@@ -0,0 +1,27 @@
+Copyright (c) 2009 The Go Authors. All rights reserved.
```
Comment (gboyvalenkov-bosch) [inline licenses/LICENSE.golang_org_x_net-0.0.0-20211123203042-d83791d6bcd9:1]: remove this file

File: licenses/LICENSE.golang_org_x_sys-0.0.0-20211124211545-fe61309f8881 [hunk 14]
```diff
@@ -0,0 +1,27 @@
+Copyright (c) 2009 The Go Authors. All rights reserved.
```
Comment (gboyvalenkov-bosch) [inline licenses/LICENSE.golang_org_x_sys-0.0.0-20211124211545-fe61309f8881:1]: remove this file

File: licenses/LICENSE.google_uuid-1.3.0 [hunk 15]
```diff
@@ -0,0 +1,27 @@
+Copyright (c) 2009,2014 Google Inc. All rights reserved.
```
Comment (gboyvalenkov-bosch) [inline licenses/LICENSE.google_uuid-1.3.0:1]: remove this file

File: licenses/LICENSE.gorilla_websocket-1.4.2 [hunk 16]
```diff
@@ -0,0 +1,22 @@
+Copyright (c) 2013 The Gorilla WebSocket Authors. All rights reserved.
```
Comment (gboyvalenkov-bosch) [inline licenses/LICENSE.gorilla_websocket-1.4.2:1]: remove this file

File: licenses/LICENSE.natefinch_lumberjack.v2-2.0.0 [hunk 17]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Comment (gboyvalenkov-bosch) [inline licenses/LICENSE.natefinch_lumberjack.v2-2.0.0:1]: remove this file

File: internal/utils_test.go [hunk 18]
```diff
@@ -88,9 +88,9 @@ func mockScriptBasedSoftwareUpdatable(t *testing.T, tc *testConfig) (*ScriptBase
 
 	// Initialize mocked ScriptBasedSoftwareUpdatable
 	if err := feature.init(&ScriptBasedSoftwareUpdatableConfig{
-		Broker:     defaultBroker,
+		Broker:     getDefaultFlagValue(flagFeatureID),
```
Comment (gboyvalenkov-bosch) [inline internal/utils_test.go:91]: check if appropriate value is used

**Review by gboyvalenkov-bosch — CHANGES_REQUESTED** [review]

File: internal/storage/download_test.go [hunk 19]
```diff
@@ -56,8 +67,59 @@ func TestDownloadToFile(t *testing.T) {
 			HashType:  "SHA256",
 			HashValue: "4eefb9a7a40a8b314b586a00f307157043c0bbe4f59fa39cba88773680758bc3",
 		},
+	}, "", "", t)
+}
+
+// TestDownloadToFileSecureSystemPool tests downloadToFile function, using secure protocol(s) and certificates from system pool.
+func TestDownloadToFileSecureSystemPool(t *testing.T) {
+	// SystemCertPool does not run on Windows:
+	//https://github.com/golang/go/issues/16736
```
Comment (e-grigorov) [inline internal/storage/download_test.go:76]: In general, the comments should use one and the same formatting i.e. there should be a space.

File: internal/storage/download_test.go [hunk 20]
```diff
@@ -210,11 +321,19 @@ func host(addr string, name string, size int64, simple bool, t *testing.T) *web
...
+	if secure {
+		go func() {
+			if err := w.srv.ListenAndServeTLS(cert, key); err != nil && err != http.ErrServerClosed {
+				t.Errorf("The secure server is not started: %v", err)
+			}
+		}()
+	} else {
+		go func() {
+			if err := w.srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+				t.Errorf("Failed to start web server: %v", err)
```
Comment (e-grigorov) [inline internal/storage/download_test.go:333]: Unify the error messages of the servers, the only one difference is the "secure".

File: internal/storage/download_test.go [hunk 21]
```diff
@@ -198,7 +309,7 @@ func check(name string, expected int, t *testing.T) {
 }
 
 // host create and start a HTTP server.
-func host(addr string, name string, size int64, simple bool, t *testing.T) *web {
+func host(addr string, name string, size int64, simple bool, secure bool, cert string, key string, t *testing.T) *web {
```
Comment (e-grigorov) [inline internal/storage/download_test.go:312]: Duplicates the "host" method in utils_test.go.

File: internal/storage/testdata/valid_key.pem [hunk 22]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (e-grigorov) [inline internal/storage/testdata/valid_key.pem:1]: - Use test certificate as it's done for "ext" and "expired"
- The validity of 1 year is not ok for test certificate

File: internal/flags.go [hunk 23]
```diff
@@ -107,59 +89,79 @@ func InitFlags(version string) (*ScriptBasedSoftwareUpdatableConfig, *logger.Log
...
+		fieldValue := valueOf.FieldByName(fieldType.Name)
+		pointer := fieldValue.Addr().Interface()
+		flagName := toFlagName(fieldType.Name)
+		switch val := fieldValue.Interface(); val.(type) {
+		case string:
+			flag.StringVar(pointer.(*string), flagName, defaultValue, description)
+		case int:
+			value, err := strconv.Atoi(defaultValue)
+			if err != nil {
+				log.Printf("error parsing integer argument %v with value %v", fieldType.Name, defaultValue)
```
Comment (e-grigorov) [inline internal/flags.go:112]: Is it the right way to handle this error i.e. there will be error log and the type default value will be set. Do we need to stop the excution?

File: internal/flags_test.go [hunk 24]
```diff
@@ -372,3 +377,25 @@ func resetArgs() {
 		flag.CommandLine = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
 	}()
 }
+
+func getDefaultFlagValue(flagName string) string {
+	flagName = replaceSuffix(flagName, "ID", "Id")
```
Comment (e-grigorov) [inline internal/flags_test.go:382]: toFieldName seems to do the same job.

File: internal/storage/download_test.go [hunk 25]
```diff
@@ -145,49 +208,97 @@ func TestDownloadToFileError(t *testing.T) {
...
+	defer srvSecureInv.close()
+	srvSecureExt := host(":43235", art.FileName, int64(art.Size), true, true, extCert, extKey, t)
+	defer srvSecureExt.close()
+	srvSecureVal := host(":43236", art.FileName, int64(art.Size), true, true, validCert, validKey, t)
+	defer srvSecureVal.close()
+	name := filepath.Join(dir, art.FileName)
+
+	// 1. Try download with expired certificate
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
+		t.Fatalf("validated with expired certificate: %v", err)
```
Comment (e-grigorov) [inline internal/storage/download_test.go:285]: The message is not absolutely correct, the client is not configured with the server certificate.

File: internal/storage/download_test.go [hunk 26]
```diff
@@ -145,49 +208,97 @@ func TestDownloadToFileError(t *testing.T) {
...
+	}
+	if err := downloadArtifact(name, art, nil, expiredCert, make(chan struct{})); err == nil {
+		t.Fatalf("validated with expired certificate: %v", err)
+	}
+
+	// 2. Try download with unauthorized certificate
+	art.Link = "https://localhost:43235/test.txt"
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
+		t.Fatalf("validated with unauthorized certificate: %v", err)
```
Comment (e-grigorov) [inline internal/storage/download_test.go:294]: "unauthorized certificate" - what does it mean here?

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: internal/flags.go (hunk 23 above)
Comment (gboyvalenkov-bosch) [reply internal/flags.go:112]: The same issue is open in file upload. Maybe they need to be resolved together.

File: internal/flags.go [hunk 27]
```diff
@@ -58,45 +64,34 @@ var (
...
+	Broker          string   `json:"broker" def:"tcp://localhost:1883" descr:"Local MQTT broker address"`
+	Username        string   `json:"username" descr:"Username for authorized local client"`
+	Password        string   `json:"password" descr:"Password for authorized local client"`
+	StorageLocation string   `json:"storageLocation" descr:"Location of the storage"`
+	FeatureID       string   `json:"featureId" def:"SoftwareUpdatable" descr:"Feature identifier of SoftwareUpdatable"`
+	ModuleType      string   `json:"moduleType" def:"software" descr:"Module type of SoftwareUpdatable"`
+	ArtifactType    string   `json:"artifactType" def:"archive" descr:"Defines the module artifact type: archive or plane"`
+	Install         []string `json:"install" descr:"Defines the absolute path to install script"`
+	ServerCert      string   `json:"serverCert" descr:"A PEM encoded certificate \"file\" for secure artifact download"`
+	LogFile         string   `json:"logFile" def:"logs/log.txt" descr:"Log file location in storage directory"`
```
Comment (e-grigorov) [inline internal/flags.go:76]: The default value is not correct, it should be: log/software-update.log

File: internal/flags.go [hunk 28]
```diff
@@ -58,45 +64,34 @@ var (
...
+	LogLevel        string   `json:"logLevel" def:"INFO" descr:"Log levels are ERROR, WARNING, INFO, DEBUG, TRACE"`
```
Comment (e-grigorov) [inline internal/flags.go:77]: Could you please double check the warning log level. According to the documentation, it's: WARN.

File: internal/storage/download.go [hunk 29]
```diff
@@ -143,6 +139,47 @@ func resume(to string, offset int64, artifact *Artifact, progress progressBytes,
...
+			logger.Errorf("Error reading CA certificate file - \"%s\"", serverCert)
+			return nil, err
+		}
+		caCertPool = x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+	}
+
+	u, _ := url.Parse(link) // MUST not return error, since http(s) request was done to that url
+	if u.Scheme == "https" {
+		config := &tls.Config{ // using the system CA pool
```
Comment (e-grigorov) [inline internal/storage/download.go:167]: "using the system CA pool" - is this comment valid? RootCAs are set and the system CA pool should not be used or I am missing something?

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: internal/flags.go (hunk 23 above)
Comment (e-grigorov) [reply internal/flags.go:125]: #15

File: internal/flags.go (hunk 27 above)
Comment (e-grigorov) [reply internal/flags.go]: It's "log" not "logs".

**Review by e-grigorov — APPROVED** [review]

**Pull request merged**